File: tests/common.tests/JsonArray.cs
using CsCheck;
using FluentAssertions;
using System.Linq;
using System.Text.Json.Nodes;
using System.Threading;
using System.Threading.Tasks;
using Xunit;

namespace common.tests;

public class JsonArrayTests
{
    [Fact]
    public void ToJsonArray_has_correct_count()
    {
        var generator = from count in Gen.Int[0, 100]
                        from nodes in Generator.JsonNode.Null().Array[count]
                        select (count, nodes);

        generator.Sample((count, nodes) =>
        {
            var jsonArray = nodes.ToJsonArray();
            jsonArray.AsEnumerable().Should().HaveCount(count);
        });
    }

    [Fact]
    public async Task ToJsonArray_with_async_enumerable_has_correct_count()
    {
        var generator = from count in Gen.Int[0, 100]
                        from nodes in Generator.JsonNode.Null().Array[count]
                        select (count, nodes.ToAsyncEnumerable());

        await generator.SampleAsync(async (count, nodes) =>
        {
            var jsonArray = await nodes.ToJsonArray(CancellationToken.None);
            jsonArray.AsEnumerable().Should().HaveCount(count);
        });
    }

    [Fact]
    public void GetJsonObjects_fails_if_the_array_contains_a_non_object_node()
    {
        var generator = from jsonArray in Generator.JsonArray
                        where jsonArray.Any(node => node is not JsonObject)
                        select jsonArray;

        generator.Sample(jsonArray =>
        {
            var result = jsonArray.GetJsonObjects();
            result.Should().BeError();
        });
    }

    [Fact]
    public void GetJsonObjects_succeeds_if_the_array_is_empty()
    {
        var jsonArray = new JsonArray();
        var result = jsonArray.GetJsonObjects();
        result.Should().BeSuccess();
    }

    [Fact]
    public void GetJsonObjects_succeeds_if_the_array_only_contains_object_nodes()
    {
        var generator = from jsonObjects in Generator.JsonObject.Array[0, 10]
                        where jsonObjects.Length > 0
                        select new JsonArray(jsonObjects);

        generator.Sample(jsonArray =>
        {
            var result = jsonArray.GetJsonObjects();
            result.Should().BeSuccess();
        });
    }

    [Fact]
    public void GetJsonArrays_fails_if_the_array_contains_a_non_array_node()
    {
        var generator = from jsonArray in Generator.JsonArray
                        where jsonArray.Any(node => node is not JsonArray)
                        select jsonArray;

        generator.Sample(jsonArray =>
        {
            var result = jsonArray.GetJsonArrays();
            result.Should().BeError();
        });
    }

    [Fact]
    public void GetJsonArrays_succeeds_if_the_array_is_empty()
    {
        var jsonArray = new JsonArray();
        var result = jsonArray.GetJsonArrays();
        result.Should().BeSuccess();
    }

    [Fact]
    public void GetJsonArrays_succeeds_if_the_array_only_contains_array_nodes()
    {
        var generator = from jsonArrays in Generator.JsonArray.Array[0, 10]
                        where jsonArrays.Length > 0
                        select new JsonArray(jsonArrays);

        generator.Sample(jsonArray =>
        {
            var result = jsonArray.GetJsonArrays();
            result.Should().BeSuccess();
        });
    }

    [Fact]
    public void GetJsonValues_fails_if_the_array_contains_a_non_value_node()
    {
        var generator = from jsonArray in Generator.JsonArray
                        where jsonArray.Any(node => node is not JsonValue)
                        select jsonArray;

        generator.Sample(jsonArray =>
        {
            var result = jsonArray.GetJsonValues();
            result.Should().BeError();
        });
    }

    [Fact]
    public void GetJsonValues_succeeds_if_the_array_is_empty()
    {
        var jsonArray = new JsonArray();
        var result = jsonArray.GetJsonValues();
        result.Should().BeSuccess();
    }

    [Fact]
    public void GetJsonValues_succeeds_if_the_array_only_contains_value_nodes()
    {
        var generator = from jsonValues in Generator.JsonValue.Array[0, 10]
                        where jsonValues.Length > 0
                        select new JsonArray(jsonValues);

        generator.Sample(jsonArray =>
        {
            var result = jsonArray.GetJsonValues();
            result.Should().BeSuccess();
        });
    }
}

File: tests/fsharp/JsonArray.fs
module JsonArrayTests

open Xunit
open common.fsharp
open Faqt
open FsCheck
open FsCheck.FSharp
open System.Text.Json.Nodes

let private generateNonEmptyJsonArray (gen: Gen<JsonArray>) =
    gen |> Gen.filter (fun jsonArray -> jsonArray.Count > 0)

[<Fact>]
let ``fromSeq has expected count`` () =
    let gen = Gen.listOf Gen.jsonNode

    Check.fromGen gen (fun nodes ->
        let jsonArray = JsonArray.fromSeq nodes
        jsonArray.Should().HaveLength(nodes.Length))

[<Fact>]
let ``asJsonObjects with non JSON objects returns a failure`` () =
    let gen =
        Gen.JsonNode.nonJsonObject |> Gen.generateJsonArray |> generateNonEmptyJsonArray

    Check.fromGen gen (fun jsonArray ->
        let result = JsonArray.asJsonObjects jsonArray
        result.Should().BeFailure())

[<Fact>]
let ``asJsonObjects with an empty array returns a success`` () =
    let jsonArray = JsonArray()
    let result = JsonArray.asJsonObjects jsonArray
    result.Should().BeSuccess().That.Should().BeEmpty()

[<Fact>]
let ``asJsonObjects with JSON objects returns a success`` () =
    let gen =
        Gen.JsonNode.jsonObject |> Gen.generateJsonArray |> generateNonEmptyJsonArray

    Check.fromGen gen (fun jsonArray ->
        let result = JsonArray.asJsonObjects jsonArray
        result.Should().BeSuccess().That.Should().HaveLength(jsonArray.Count))

[<Fact>]
let ``asJsonArrays with non JSON arrays returns a failure`` () =
    let gen =
        Gen.JsonNode.nonJsonArray |> Gen.generateJsonArray |> generateNonEmptyJsonArray

    Check.fromGen gen (fun jsonArray ->
        let result = JsonArray.asJsonArrays jsonArray
        result.Should().BeFailure())

[<Fact>]
let ``asJsonArrays with an empty array returns a success`` () =
    let jsonArray = JsonArray()
    let result = JsonArray.asJsonArrays jsonArray
    result.Should().BeSuccess().That.Should().BeEmpty()

[<Fact>]
let ``asJsonArrays with JSON arrays returns a success`` () =
    let gen =
        Gen.JsonNode.jsonArray |> Gen.generateJsonArray |> generateNonEmptyJsonArray

    Check.fromGen gen (fun jsonArray ->
        let result = JsonArray.asJsonArrays jsonArray
        result.Should().BeSuccess().That.Should().HaveLength(jsonArray.Count))

[<Fact>]
let ``asJsonValues with non JSON values returns a failure`` () =
    let gen =
        Gen.JsonNode.nonJsonValue |> Gen.generateJsonArray |> generateNonEmptyJsonArray

    Check.fromGen gen (fun jsonArray ->
        let result = JsonArray.asJsonValues jsonArray
        result.Should().BeFailure())

[<Fact>]
let ``asJsonValues with an empty array returns a success`` () =
    let jsonArray = JsonArray()
    let result = JsonArray.asJsonValues jsonArray
    result.Should().BeSuccess().That.Should().BeEmpty()

[<Fact>]
let ``asJsonValues with JSON values returns a success`` () =
    let gen =
        Gen.JsonNode.jsonValue |> Gen.generateJsonArray |> generateNonEmptyJsonArray

    Check.fromGen gen (fun jsonArray ->
        let result = JsonArray.asJsonValues jsonArray
        result.Should().BeSuccess().That.Should().HaveLength(jsonArray.Count))

[<Fact>]
let ``getJsonObjects returns only JSON objects`` () =
    let gen =
        gen {
            let! jsonObjects = Gen.jsonObject |> Gen.arrayOf

            let! jsonArray =
                let jsonObjectsAsNodes =
                    jsonObjects |> Array.map (fun jsonObject -> jsonObject :> JsonNode)

                Gen.JsonNode.nonJsonObject
                |> Gen.arrayOf
                |> Gen.map (Array.append jsonObjectsAsNodes)
                |> Gen.bind Gen.shuffle
                |> Gen.map JsonArray.fromSeq

            return jsonObjects, jsonArray
        }

    Check.fromGen gen (fun (jsonObjects, jsonArray) ->
        let result = JsonArray.getJsonObjects jsonArray
        result.Should().HaveSameCountAs(jsonObjects))

[<Fact>]
let ``getJsonArrays returns only JSON arrays`` () =
    let gen =
        gen {
            let! jsonArrays = Gen.jsonArray |> Gen.arrayOf

            let! jsonArray =
                let jsonArraysAsNodes =
                    jsonArrays |> Array.map (fun jsonArray -> jsonArray :> JsonNode)

                Gen.JsonNode.nonJsonArray
                |> Gen.arrayOf
                |> Gen.map (Array.append jsonArraysAsNodes)
                |> Gen.bind Gen.shuffle
                |> Gen.map JsonArray.fromSeq

            return jsonArrays, jsonArray
        }

    Check.fromGen gen (fun (jsonArrays, jsonArray) ->
        let result = JsonArray.getJsonArrays jsonArray
        result.Should().HaveSameCountAs(jsonArrays))

[<Fact>]
let ``getJsonValues returns only JSON values`` () =
    let gen =
        gen {
            let! jsonValues = Gen.jsonValue |> Gen.arrayOf

            let! jsonArray =
                let jsonValuesAsNodes =
                    jsonValues |> Array.map (fun jsonValue -> jsonValue :> JsonNode)

                Gen.JsonNode.nonJsonValue
                |> Gen.arrayOf
                |> Gen.map (Array.append jsonValuesAsNodes)
                |> Gen.bind Gen.shuffle
                |> Gen.map JsonArray.fromSeq

            return jsonValues, jsonArray
        }

    Check.fromGen gen (fun (jsonValues, jsonArray) ->
        let result = JsonArray.getJsonValues jsonArray
        result.Should().HaveSameCountAs(jsonValues))
